Implement newAzureCredential function for Azure authentication

priyansh17 opened the pull request:
Implement newAzureCredential function for Azure authentication and refactor credential retrieval:
Documentation: https://learn.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet

Fixes #(issue)
None


- [X] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.

priyansh17 commented:
/kind changelog-not-required

kaovilai commented:
Can you walk me through any new added ways to auth, or if any auth breaking changes are anticipated?

priyansh17 commented:
> Can you walk me through any new added ways to auth, or if any auth breaking changes are anticipated?

Hi, no they are no breaking changes. I have just added the fallback mechanism if the certs or credentials are not provided it uses other methods as mentioned in the doc. this new sdk internally uses all other methods like workload identity, cli creds etc. I have added the link to methods sdk uses in the description.

priyansh17 commented:
> Is there any docs needing update in azure plugin?

No, I think it will still work the same way there is no change-log needed or doc updation required. If you feel some change needs to be documented for test framework please let me know. This is updated only in e2e tests.

For context I am developing a pipeline to run azure plugin tests where I was facing auth issues as we don't allow service principals or certs now. This is where I am making changes only in e2e frameworks similar to other previous PRs related to Kibishii.

kaovilai commented:
> where I was facing auth issues as we don't allow service principals or certs now

what about non test scenarios, would the same requirement apply and is it already working such that only test needed this update?

priyansh17 commented:
> > where I was facing auth issues as we don't allow service principals or certs now
> 
> what about non test scenarios, would the same requirement apply and is it already working such that only test needed this update?

In non test scenarios we can leave it as it is for now, I don't see an issue in those cases. Outside Azure customers are still free to create certs and service principals but for our internal org we would need to raise exceptions, those get flagged and such issues etc.
So to prevent all that in tests I have added a new sdk along with the older one in place. If older one fails, it falls back to this. 

Hence only tests need an update for now.
